src/relay/pass/type_solver.cc

```diff
@@ -102,35 +149,65 @@ class TypeSolver::Unifier : public TypeFunctor<Type(const Type&, const Type&)> {
     if (lhs->FindRoot() == rhs->FindRoot()) {
       return lhs->resolved_type;
     }
+
+    std::cout << "LHS: " << lhs->resolved_type << " hints " << lhs->FindRoot()->hint << std::endl;
+    std::cout << "RHS: " << rhs->resolved_type << " hints " << rhs->FindRoot()->hint << std::endl;
+
     if (lhs->resolved_type.as<IncompleteTypeNode>()) {
-      CHECK(!CheckOccurs(lhs, rhs->resolved_type))
+      CHECK(!OccursCheck(lhs, rhs->resolved_type))
         << "Incomplete type " << lhs->resolved_type << " occurs in "
         << rhs->resolved_type << ", cannot unify";
+
       solver_->MergeFromTo(lhs, rhs);
       return rhs->resolved_type;
     } else if (rhs->resolved_type.as<IncompleteTypeNode>()) {
-      CHECK(!CheckOccurs(rhs, lhs->resolved_type))
+      CHECK(!OccursCheck(rhs, lhs->resolved_type))
         << "Incomplete type " << rhs->resolved_type << " occurs in "
         << lhs->resolved_type << ", cannot unify";
       solver_->MergeFromTo(rhs, lhs);
       return lhs->resolved_type;
     } else {
-      Type resolved = this->VisitType(lhs->resolved_type, rhs->resolved_type);
-      CHECK(resolved.defined())
-        << "Unable to unify parent types: "
-        << lhs->resolved_type << " and " << rhs->resolved_type;
-      TypeNode* top = solver_->GetTypeNode(resolved);
-      solver_->MergeFromTo(lhs, top);
-      solver_->MergeFromTo(rhs, top);
-      return resolved;
+      if (lhs->hint.defined() || rhs->hint.defined()) {
```

Comment:
It would be good to have some explanation of what this is doing at a high level: If there are hints for tensor types we are unifying, we use the hints to produce a generalized tensor type. It would be good to assert that the types must be tensor types, too.